model/Command/ChangePermissionsCommand.php

```diff
@@ -0,0 +1,117 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2023 (original work) Open Assessment Technologies SA.
+ */
+
+declare(strict_types=1);
+
+namespace oat\taoDacSimple\model\Command;
+
+use core_kernel_classes_Resource;
+
+/**
+ * Value object holding data about permission changes to be done in a resource.
+ *
+ * By default, the command is set up to apply a given set of ACLs to a single
+ * resource designated as the command's root.
+ *
+ * However, it also provides methods to create copies of the command having
+ * different recursion settings, making PermissionsService to operate
+ * recursively or to modify class instances under the root resource in case the
+ * root itself is a class.
+ */
+class ChangePermissionsCommand
+{
+    private core_kernel_classes_Resource $root;
+
+    /**
+     * An array in the form ['userId' => ['READ', 'WRITE], ...]
+     *
+     * @var string[]
+     */
+    private array $privilegesPerUser;
+
+    private bool $isRecursive = false;
+
+    private bool $applyToNestedResources = false;
+
+    public function __construct(
+        core_kernel_classes_Resource $root,
+        array $privileges
+    ) {
+        $this->root = $root;
+        $this->privilegesPerUser = $privileges;
+    }
+
+    /**
+     * Clones the current command setting its recursive flag to a new value.
+     *
+     * For commands having a class as the root, setting the recursion flag
+     * makes PermissionsService to update permissions for the class and all its
+     * descendants (i.e. updates all resources AND classes using the provided
+     * root class as the initial node for a tree traversal, which may be slow
+     * and resource-intensive).
+     *
+     * This is provided for backward compatibility purposes.
+     */
+    public function withRecursion(bool $isRecursive = true): self
+    {
+        $ret = clone $this;
+        $ret->isRecursive = $isRecursive;
+
+        return $ret;
+    }
+
+    /**
+     * Clones the current command setting its applyToNestedResources flag to a
+     * new value.
+     *
+     * For commands having a class as the root AND not having the recursion
+     * flag set, setting the nested resources flag makes PermissionsService to
+     * update permissions for the class and all instances of that class, but
+     * skips all nested classes and instances of them (i.e. does not go down
+     * into nested levels of the resource tree).
+     */
+    public function withNestedResources(bool $applyToNestedResources = true): self
+    {
+        $ret = clone $this;
+        $ret->applyToNestedResources = $applyToNestedResources;
+
+        return $ret;
```

Comment:
We do not need to clone a command. 

```suggestion
        $ret->applyToNestedResources = $applyToNestedResources;

        return $this;
```

Reply:
I was modeling the command as a value object (that is, an immutable object, so further changes need to return new instances). I'll change that.

Reply:
IMO it does not justify increasing memory consumption because of this. For what we have now, there is no use case where we want to generate different instances of commands from one Command state.

Reply:
1abe4f4f36035340159f4a50e8627d48ae5e9af4